### PR TITLE
[Swift] Add tuple concept exercise

### PR DESCRIPTION
**Pull request by wneumann**
Added the tuple concept exercise.

I am a little concerned about the story as it is based around the Santa Claus story, which plays better in some countries/cultures than others.

### CODE REVIEW

File: languages/swift/exercises/concept/tuples/.docs/instructions.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+You are an elf working in Santa Claus' logistics division and you have been given a pair of tasks from the boss for upgrading the system's software.
```
Comment (stillleben) [inline languages/swift/exercises/concept/tuples/.docs/instructions.md:1]: I missed the debate about Santa Claus (I understand, that it might not be familiar to all students) - and maybe I'm not aware of some common agreement regarding this topic. 

My personal thoughts by reading these 2 exercises: It is perfectly doable without knowing Santa Claus ... it is then just "some business". :)

File: languages/swift/exercises/concept/tuples/.docs/hints.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+## General
+
+The individual components of a tuple can be accessed by appending a _.n_ to the name of the tuple, where _n_ is the index of the component, statring with 0.
```
Comment (stillleben) [inline languages/swift/exercises/concept/tuples/.docs/hints.md:3]:
```suggestion
The individual components of a tuple can be accessed by appending a _.n_ to the name of the tuple, where _n_ is the index of the component, starting with 0.
```

File: languages/swift/exercises/concept/tuples/.docs/hints.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+## 1. Convert coordinates to polar
+
+- The individual components of a labeled tuple can also be accessed by appending a _.label_ to the name of the tuple, where _label_ is the label name of the component.
+- The `atan2(_:,_:)`function requires `Foundation` to be imported.
+- Floating point types in Swift have the [squareRoot() method][squareroot-docs].
+
+## 2. Merge two database records
+
+- Tuples can be [decomposed and assigned to individual constants or variables][tuples].
```
Comment (stillleben) [inline languages/swift/exercises/concept/tuples/.docs/hints.md:13]: From the student's perspective I find the link to the general tuples-documentation a little confusing because I need to grasp these examples and might have problems to find or understand it.
What about an inline example for decomposing a tuple from the introduction?

File: languages/swift/exercises/concept/tuples/.docs/after.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+```swift
+let order: (String, Double, Int) = ("biscuits", 2.99, 3)
+```
+
+The fact that tuples are single values allows them to be used to [return multiple values from a function][multiple-return-values]:
+
+```swift
+func longestAndShortest(_ strings: [String]) -> (longest: String?, shortest: String?) {
+  var longestLength = Int.min, shortestLength = Int.max
+  var longAndShort: (longest: String?, shortest: String?) = (longest: nil, shortest: nil)
```
Comment (stillleben) [inline languages/swift/exercises/concept/tuples/.docs/after.md:12]: I'm thinking about that implementation. What about using two separate strings `longest` and `shortest` and to construct the tuple only in return statement? That might also underline the idea of using tuples to pass multiple return values (while within the function and outside it can be worked with the separate variables).

File: languages/swift/exercises/concept/tuples/.docs/after.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    }
+  }
+  return longAndShort
+}
+
+longestAndShortest(["robin", "blue jay", "heron", "swift", "swan"])
+// => (longest: "blue jay", shortest: "swan")
+```
+
+The individual items of a tuple may be accessed by appending a `.n` to the tuple's name where _n_ is the index of the element you would like to access, starting with 0.
```
Comment (stillleben) [inline languages/swift/exercises/concept/tuples/.docs/after.md:30]: As here starts a "recap" (what has been learned from introduction and the exercise) what about inserting a `Key Takeaways from this Exercise` heading here like in https://github.com/exercism/v3/blob/master/languages/swift/exercises/concept/dictionaries/.docs/after.md#key-takeaways-from-this-exercise?

File: languages/swift/exercises/concept/tuples/.meta/Example.swift [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import Foundation
+
+func cartesianToPolar(_ cart: (x: Double, y: Double)) -> (r: Double, theta: Double) {
+  (r: (cart.x * cart.x + cart.y * cart.y).squareRoot(), theta: atan2(cart.y, cart.x))
```
Comment (stillleben) [inline languages/swift/exercises/concept/tuples/.meta/Example.swift:4]: Is it possible in Swift to omit the `return` keyword? (would be another similarity to Scala) ;)

File: languages/swift/exercises/concept/tuples/.meta/Example.swift [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+}
+
+func combineRecords(
+  production: (toy: String, id: Int, productLead: String),
+  gifts: (Int, [String])
+) -> (id: Int, toy: String, productLead: String, recipients: [String]) {
+  return (id: production.id, toy: production.toy, productLead: production.productLead, gifts.1)
```
Comment (stillleben) [inline languages/swift/exercises/concept/tuples/.meta/Example.swift:11]: see my comment in instruction.md: Maybe add some handling of different ids here.

File: languages/swift/exercises/concept/tuples/.meta/design.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+## Resources to refer to
+
+### Hints
+
+- [Tuples][tuples]
+
+### After
+
+- [Tuples][tuples]
+- [Functiuons with multiple return values][multiple-return-values]
```
Comment (stillleben) [inline languages/swift/exercises/concept/tuples/.meta/design.md:33]:
```suggestion
- [Functions with multiple return values][multiple-return-values]
```

File: languages/swift/exercises/concept/tuples/README.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
+You are an elf working in Santa Claus' logistics division and you have been given a pair of tasks from the boss for upgrading the system's software.
```
Comment (stillleben) [inline languages/swift/exercises/concept/tuples/README.md:1]: Please see my comments on `instructions.md`.

File: languages/swift/exercises/concept/tuples/Tests/tuplesTests/tuplesTests.swift [hunk 10]
```diff
@@ -0,0 +1,61 @@
+import XCTest
+
+@testable import tuples
+
+final class tuplesTests: XCTestCase {
+  private let runAll =
+    Bool(ProcessInfo.processInfo.environment["RUNALL", default: "false"]) ?? false
+  private let aboutEqual = { abs($0 - $1) < 0.001 }
+
+  func testCaretesianToPolar() {
```
Comment (stillleben) [inline languages/swift/exercises/concept/tuples/Tests/tuplesTests/tuplesTests.swift:10]:
```suggestion
  func testCartesianToPolar() {
```

File: languages/swift/exercises/concept/tuples/Tests/tuplesTests/tuplesTests.swift [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    let coordinate = (x: 11.713, y: 5.6405)
+    let (actualR, actualTheta) = cartesianToPolar(coordinate)
+    XCTAssertTrue(aboutEqual(actualR, 13.0) && aboutEqual(actualTheta, 0.4488))
+  }
+
+  func testCaretesianToPolarQ3() throws {
```
Comment (stillleben) [inline languages/swift/exercises/concept/tuples/Tests/tuplesTests/tuplesTests.swift:16]:
```suggestion
  func testCartesianToPolarQ3() throws {
```

File: languages/swift/exercises/concept/tuples/Tests/tuplesTests/tuplesTests.swift [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+  //    func testPreperationMinutes() throws {
+  //        try XCTSkipIf(true && !runAll) // change true to false to run this test
+  //        XCTAssertEqual(preparationTimeInMinutes(layers: 6), 12)
+  //    }
+
+  static var allTests = [
+    ("testCaretesianToPolar", testCaretesianToPolar),
+    ("testCaretesianToPolarQ3", testCaretesianToPolarQ3),
+    ("testcombineRecords", testcombineRecords),
```
Comment (stillleben) [inline languages/swift/exercises/concept/tuples/Tests/tuplesTests/tuplesTests.swift:59]:
```suggestion
    ("testCartesianToPolar", testCartesianToPolar),
    ("testCartesianToPolarQ3", testCartesianToPolarQ3),
    ("testcombineRecords", testcombineRecords),
```

File: languages/swift/exercises/concept/tuples/Tests/tuplesTests/tuplesTests.swift [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+      actual.id == id && actual.toy == toy && actual.productLead == productLead
+        && actual.recipients == recipients
+    )
+  }
+
+  //    func testPreperationMinutes() throws {
+  //        try XCTSkipIf(true && !runAll) // change true to false to run this test
+  //        XCTAssertEqual(preparationTimeInMinutes(layers: 6), 12)
+  //    }
+
```
Comment (stillleben) [inline languages/swift/exercises/concept/tuples/Tests/tuplesTests/tuplesTests.swift:55]: Can maybe be thrown out?

File: languages/swift/exercises/concept/tuples/Sources/tuples/tuples.swift [hunk 14]
```diff
@@ -0,0 +1,12 @@
+import Foundation
+
+func cartesianToPolar(_ cart: (x: Double, y: Double)) -> (r: Double, theta: Double) {
+  fatalError("Please implement the cartesianToPolar(_:) function.")
+}
+
+func combineRecords(
+  production: (toy: String, id: Int, prouctLead: String),
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/tuples/Sources/tuples/tuples.swift:8]: Strings are used here, but I don't see them as a prerequisite in the `config.json` file

File: languages/swift/exercises/concept/tuples/Sources/tuples/tuples.swift [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+  gifts: (Int, [String])
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/tuples/Sources/tuples/tuples.swift:9]: Arrays are used here, but I don't see them as a prerequisite in the `config.json` file

File: languages/swift/exercises/concept/tuples/.docs/instructions.md [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+## 1. Convert coordinates to polar
+
+Your first task is updating the complex number array that helps run Santa's sleigh. The boss discovered that the last development team stored the arrays using Cartesian coordinates rather than polar coordinates, which everyone knows are the boss' favorite.
+
+For this task you will need to implement the function `cartesianToPolar(_:)` that takes in a tuple, `(x: Double, y: Double)`, and return a new tuple, `(r: Double, theta: Double)`, where:
+
+- r = √(x<sup>2</sup> + y<sup>2</sup>)
+- theta = _atan2_(y, x)
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/tuples/.docs/instructions.md:10]: @iHiD What do you think about this math-based task? It is fairly light on math, but I'd appreciate a second look :)

File: languages/swift/exercises/concept/tuples/.docs/instructions.md [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+    toy: "Chemistry set",
+    id: 328509,
+    productLead: "Binkles"
+ ),
+  gifts: (
+    328509,
+    ["Inés", "Maxime", "Bandile", "Shaurya", "Екатерина"]
+  )
+)
+// ==> (id: 328509, toy: "Chemistry set", productLead: "Binkles", recipients: ["Inés", "Maxime", "Bandile", "Shaurya", "Екатерина"])
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/tuples/.docs/instructions.md:36]:
```suggestion
// => (id: 328509, toy: "Chemistry set", productLead: "Binkles", recipients: ["Inés", "Maxime", "Bandile", "Shaurya", "Екатерина"])
```

File: languages/swift/exercises/concept/tuples/.docs/instructions.md [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+```
```
[the remark this answers is not in the document]
Comment (ErikSchierboom) [reply languages/swift/exercises/concept/tuples/.docs/instructions.md:37]: > Could you specify, what should happen, if given ids from both tuples don't match, e.g. with a second example?

I would rather not add this, as in general we'd like to keep error handling out of concept exercises, unless the concept is error handling :)

File: languages/swift/exercises/concept/tuples/.docs/instructions.md (hunk 1 above)
Comment (ErikSchierboom) [reply languages/swift/exercises/concept/tuples/.docs/instructions.md:1]: I quite like the Santa Claus theme :) It gives it a bit of flourish. 

> I missed the debate about Santa Claus

What was the debate?

File: languages/swift/exercises/concept/tuples/.docs/introduction.md [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+
+If the tuple is defined as a variable, the individual components can be updated by assigning new values to their labels or indices.
+
+```swift
+var birdSightingLog = (week: 28, birds: ["robin", "blue jay", "heron"])
+birdSightingLog.0 = 29
+// => (week 29, ["robin", "blue jay", "heron"])
+birdSightingLog.birds.append("swift")
+// => (week 29, ["robin", "blue jay", "heron", "swift"])
+```
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/tuples/.docs/introduction.md:50]: It looks like this information is not needed to solve the exercise? If so, it can be removed here and mentioned only in the `after.md` document.

File: languages/swift/exercises/concept/tuples/.docs/introduction.md [hunk 20]
```diff
@@ -0,0 +1,50 @@
+Tuples are used to combine multiple values into a single compound value where each of the values may have a different type. The compound value has a type that is the combination of the individual values' types.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/tuples/.docs/introduction.md:1]: This is a very nice introduction! You've managed to keep it quite short 👍

File: languages/swift/exercises/concept/tuples/.docs/hints.md (hunk 3 above)
Comment (ErikSchierboom) [reply languages/swift/exercises/concept/tuples/.docs/hints.md:13]: This is generally not how our hints are structured. This mention alone should be enough for the student to re-check the introduction. I also found the general documentation quite well-written, but that's personal of course.

File: languages/swift/exercises/concept/tuples/.docs/hints.md [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+- If only some of the tuple’s values are needed, unwanted components can be ignored by using an underscore (`_`) for the variable name when you decompose the tuple.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/tuples/.docs/hints.md:14]:
```suggestion
- If only some of the tuple's values are needed, unwanted components can be ignored by using an underscore (`_`) for the variable name when you decompose the tuple.
```

File: languages/swift/exercises/concept/tuples/.docs/hints.md [hunk 22]
```diff
@@ -0,0 +1,17 @@
+## General
+
+- The individual components of a tuple can be accessed by appending a _.n_ to the name of the tuple, where _n_ is the index of the component, starting with 0.
+
+## 1. Convert coordinates to polar
+
+- The individual components of a labeled tuple can also be accessed by appending a _.label_ to the name of the tuple, where _label_ is the label name of the component.
+- The `atan2(_:,_:)`function requires `Foundation` to be imported.
+- Floating point types in Swift have the [squareRoot() method][squareroot-docs].
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/tuples/.docs/hints.md:9]: Could you maybe rephrase a bit not to give away the exact solution (as in not mentioning `atan2` or `squareRoot`)?

File: languages/swift/exercises/concept/tuples/.docs/after.md (hunk 5 above)
Comment (ErikSchierboom) [reply languages/swift/exercises/concept/tuples/.docs/after.md:30]: I think that exercise is the only one that uses that format, the rest uses this format.

**Review by ErikSchierboom — APPROVED** [review]
Nice!

File: languages/swift/exercises/concept/tuples/.docs/instructions.md (hunk 1 above)
Comment (ErikSchierboom) [reply languages/swift/exercises/concept/tuples/.docs/instructions.md:1]: @iHiD Are you aware of any issues with using Santa Claus as the story?

File: languages/swift/exercises/concept/tuples/Sources/tuples/tuples.swift (hunk 15 above)
Comment (wneumann) [reply languages/swift/exercises/concept/tuples/Sources/tuples/tuples.swift:9]: Adding strings and arrays.

File: languages/swift/exercises/concept/tuples/.docs/introduction.md (hunk 19 above)
Comment (wneumann) [reply languages/swift/exercises/concept/tuples/.docs/introduction.md:50]: 👍

File: languages/swift/exercises/concept/tuples/.docs/instructions.md (hunk 1 above)
Comment (iHiD) [reply languages/swift/exercises/concept/tuples/.docs/instructions.md:1]: Nein. We're good.

File: languages/swift/exercises/concept/tuples/.docs/instructions.md (hunk 16 above)
Comment (iHiD) [reply languages/swift/exercises/concept/tuples/.docs/instructions.md:10]: - Could we use more meaningful variable names that aren't `r` and `theta`?
- Have we taught how to do sqroot somewhere?
- I don't know what `atan` is so that needs explaining if its not.
- cc @SaschaMann and @TheLostLambda for thoughts and mathsjax or whatever its called that makes formulas look pretty.

Comment (SaschaMann) [reply languages/swift/exercises/concept/tuples/.docs/instructions.md:10]:
```suggestion
- $r = \sqrt{x^2+y^2}$
- $\vartheta = \operatorname{atan2}(y, x)$
```

I'm not sure about the delimiters that would be used (since I don't know how the markdown is translated to HTML), but I used `$` and `$$` as placeholders until we can test it. This is rendered as follows:
grafik

---

Side note: I'd suggest adding the definition of `atan2` somewhere to the instructions, I had to look it up. You can copy the LaTeX source from the Wikipedia source here if you want to save time: https://en.wikipedia.org/wiki/Polar_coordinate_system#Converting_between_polar_and_Cartesian_coordinates

> Could we use more meaningful variable names that aren't r and theta?

In my unsolicited opinion: in this context, any other variable names would be less meaningful.

Comment (iHiD) [reply languages/swift/exercises/concept/tuples/.docs/instructions.md:10]: It was not obvious to me that `r` meant `radius` and I've never come across `theta` in my life before. 

Based on the fact I'm pretty well educated (with a Comp Sci degree) I would contest that most people that haven't studied maths to some more significant (and/or recent) extent would remember or have come in contact with this 🙂

Comment (SaschaMann) [reply languages/swift/exercises/concept/tuples/.docs/instructions.md:10]: Sorry, let me clarify. Those who have seen polar coordinates or similar geometric constructs before will expect those names. For those that haven't, either the name itself doesn't matter, they can accept it as it is and just plug it into the formula, or, if they want to fully understand the idea behind them, they will benefit from the canonical names because all resources will use them. If we use something else they'd have to translate between the exercise intro and the 3b1b video they're watching. For the same reason we call a tuple a tuple and not a collectionOfElementsWithDifferentTypes, even if the name tuple is meaningless to some people.

Comment (ErikSchierboom) [reply languages/swift/exercises/concept/tuples/.docs/instructions.md:10]: One possible solution would be to mention somewhere that `r` is the radius.

**Review by ErikSchierboom — APPROVED** [review]
I've approved. The [discussion on the formula](https://github.com/exercism/v3/pull/1820/files#r479080842) should be resolved either in this PR or a follow-up PR.

File: languages/swift/exercises/concept/tuples/.docs/instructions.md (hunk 16 above)
Comment (SaschaMann) [reply languages/swift/exercises/concept/tuples/.docs/instructions.md:10]: Or include a picture: https://en.wikipedia.org/wiki/Polar_coordinate_system#/media/File:Polar_to_cartesian.svg

Comment (iHiD) [reply languages/swift/exercises/concept/tuples/.docs/instructions.md]: @SaschaMann :+1:
